truffleruby/src/main/ruby/core/truffle/string_ops.rb

```diff
@@ -0,0 +1,111 @@
+module Truffle
+  class StringOps
+
+    def self.gsub_internal(orig, pattern, replacement=undefined, &block)
+      # Because of the behavior of $~, this is duplicated from gsub! because
+      # if we call gsub! from gsub, the last_match can't be updated properly.
+
+      unless orig.valid_encoding?
+        raise ArgumentError, "invalid byte sequence in #{orig.encoding}"
+      end
+
+      if undefined.equal? replacement
+        use_yield = true
+        tainted = false
+      else
+        tainted = replacement.tainted?
+        untrusted = replacement.untrusted?
+
+        unless replacement.kind_of?(String)
+          hash = Rubinius::Type.check_convert_type(replacement, Hash, :to_hash)
+          replacement = StringValue(replacement) unless hash
+          tainted ||= replacement.tainted?
+          untrusted ||= replacement.untrusted?
+        end
+        use_yield = false
+      end
+
+      pattern = Rubinius::Type.coerce_to_regexp(pattern, true) unless pattern.kind_of? Regexp
+      match = pattern.search_region(orig, 0, orig.bytesize, true)
+
+      Regexp.set_block_last_match(block, match) if block_given?
```

Comment:
It's a bit counter-intuitive this method can still set the last match, maybe add a comment here about it as it's just more convenient?